### DISCUSSION TITLE
awc websocket and actix_web_actors

### DISCUSSION THREAD

**Discussion by alaravel** [actix/actix-web#3414] [Q&A]
Hi everyone
How can pass framed to ws::start?
It's  expected Result<Bytes, PayloadError> but framed is  Result<Frame, ...>
How I can convert Frame to Bytes? 

```rust
let con = awc::Client::new().ws(url).connect().await;
match con {
    Ok((mut res, mut framed)) => {
        let req = TestRequest::default()
            .insert_header((
                "sec-websocket-key",
                res.headers().get("sec-websocket-accept").unwrap(),
            ))
            .insert_header(("sec-websocket-version", "13"))
            .insert_header(("connection", "upgrade"))
            .insert_header(("upgrade", "websocket"))
            .to_http_request();

        match actix_web_actors::ws::start(
            WsSession::new(id, room, addr),
            &req,
            framed, //todo
        ) {
            Ok(a) => {}
            Err(e) => {
                println!("error ={:?}", e)
            }
        }
    }
    Err(_) => {}
}
```

**Comment by Samielakkad** [comment]
`awc::Client::new().ws(url).connect().await` and `actix_web_actors::ws::start(...)` are for opposite sides of the WebSocket flow, so the types do not line up on purpose.

`ws::start(actor, &req, stream)` is for accepting an incoming WebSocket upgrade on the server side. The third argument is the original HTTP request payload stream, before it has been upgraded and decoded into WebSocket frames.

But `awc` gives you this after connecting as a client:

```rust
Ok((response, framed))
```

That `framed` value is already a WebSocket framed connection. Its items are WebSocket `Frame`s/messages, not raw HTTP request body bytes. You generally should not try to convert `Frame` back into `Bytes` for `ws::start`; that would be going backwards in the protocol stack.

If your goal is to write a WebSocket client, use the `framed` sink/stream directly:

```rust
use futures_util::{SinkExt as _, StreamExt as _};

let (_res, mut framed) = awc::Client::new().ws(url).connect().await?;

framed
    .send(awc::ws::Message::Text("hello".into()))
    .await?;

while let Some(item) = framed.next().await {
    match item? {
        awc::ws::Frame::Text(bytes) => {
            // handle incoming text frame
        }
        awc::ws::Frame::Binary(bytes) => {
            // handle incoming binary frame
        }
        _ => {}
    }
}
```

If your goal is to proxy/bridge between an inbound Actix websocket actor and an outbound `awc` websocket connection, then the shape is different:

- accept the browser/client connection with `actix_web_actors::ws::start`;
- create the outbound `awc` websocket connection separately;
- forward inbound actor messages to the outbound sink;
- forward outbound `awc` frames back into the actor using an actor message or `ctx.add_stream(...)`;
- in `StreamHandler`, convert those incoming frames into `ctx.text(...)`, `ctx.binary(...)`, etc.

So the short answer is: there is no useful `Frame -> Payload` conversion for `ws::start`. Use `ws::start` only for accepting server-side upgrade requests. Use the `awc` framed connection directly when you are acting as a WebSocket client.
